# Project 1 Report

### Question 1

Overview: How many rows (samples) and columns (features) are present in the dataset?

In [1]:
import pandas as pd
df=pd.read_csv('Project1-Classification.csv')
df.shape

(3150, 8)

##### There are 3150 rows (samples) and 8 columns (features) in the data set. 

Interpret Plots: Provide qualitative interpretations of the histograms.

##### The alpha-numeric histogram right skewed with an average around 2,500 characters. Additionally, most alpha-numeric character counts lie within 0 and 10,000.
##### The leaf_label histogram is evenly distributed as all categories have the same frequency, which is 350.
##### The root_label histogram is not evenly distributed. The sports category appears more frequently than the climate category; sports appears 1750 times while climate appears 1400 times. 

### Question 2

Report the number of training and testing samples

##### There are 2520 training samples and 630 testing samples

### Question 3

What are the pros and cons of lemmatization versus stemming? How do these processes affect the dictionary size?

##### Lemmatizing considers the morphology of the word being analysed; however, it requires the words correct part of speech to accurately lemmatize it. For example, the word `dusting` requires it to be identified as a verb to correctly result in the lemma  `dust`. If it is identified as a noun (as it is the default part of speech when using the lemmatize function ), the lemma `dusting` will be returned. On the other hand, stemming only removes the end portions of a word to try to find its stem. However, it maybe inaccurately stem a word. For example, stemming the word `carries` may result in the word `car` or `carri`, which are both incorrect. 
##### Both lemmatization and stemming will reduce the dictionary size as only the lemma or stems of words are kept after the process is applied. For this same reason, it will also help to reduce the run time of the algorithm as less words are processed by the algorithm. If all words in a text are already in their lemma or stem form, it will not change the dictionary size. 

min_df means minimum document frequency. How does varying min_df change the TF-IDF matrix?

##### When using CountVectorizer, min_df is used to filter out words that have a frequency below min_df. Increasing this value will filter out more words and thus decrease the TF-IDF matrix columns. Decreasing the value will increase the TF-IDF matrix columns. This change in the matrix can affect the results of the analysis being done. 

 
Should I remove stopwords before or after lemmatizing? Should I remove punctuations before or after lemmatizing? Should I remove numbers before or after lemmatizing? Hint: Recall that the full sentence is input into the Lemmatizer and the lemmatizer is tagging the position of every word based on the sentence structure.

##### Punctuations and numbers should be removed before lemmatizing as words with punctuations included are treated differently than the same word without punctuation, such as `fun!` and `fun`. Additionally, numbers are not as important when analysing the text. Removing stopwords before or after lemmatizing depends what the task requires. In cases where they may not be very useful or necessary (like sentiment analysis), they should be removed before lemmatizing. To determine whether lemmatizing should come before or after stopword removal, the intent of the task should be examined to see if they are necessary. Additionally, the task can be done with both methods to determine whether it is better the lemmatize before or after stopword removal. Additionally, the removal of these elements will reduce the run time of the algorithm analysing the text as there are less words that must be processed. 

Report the shape of the TF-IDF-processed train and test matrices. The number of rows should match the results of Question 2. The number of columns should roughly be in the order of k×103. This dimension will vary depending on your exact method of cleaning and lemmatizing and that is okay.

##### The shape of the train matrix is (2520, 14465) and the shape of the test matrix is (630, 14465); for both, the number of rows match the results from section 2. 
